base/slicearray.jl

```diff
@@ -0,0 +1,222 @@
+3×1 Slices{2, (1, Colon()), Matrix{Int64}, CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
+```
+"""
+@inline function eachslice(A; dims, drop=true)
+    _eachslice(A, dims, drop)
+end
+
+"""
+    eachrow(A::AbstractVecOrMat)
+
+Create a [`Rows`](@ref) that indexes over the rows of a vector or matrix `A`.
+
+See also [`eachcol`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachrow(a)
+2-element Rows{Matrix{Int64}, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2]
+ [3, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+```
+"""
+eachrow(A::AbstractMatrix) = _eachslice(A, (1,), true)
+eachrow(A::AbstractVector) = eachrow(reshape(A, size(A,1), 1))
+
+"""
+    eachcol(A::AbstractVecOrMat)
+
+Create a [`Columns`](@ref) that iterates over the second dimension of matrix `A`, returning the
+columns as `AbstractVector` views.
+
+See also [`eachrow`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachcol(a)
+2-element Columns{Matrix{Int64}, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}}:
+ [1, 3]
+ [2, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, :, 1) with eltype Int64:
+ 1
+ 3
+```
+"""
+eachcol(A::AbstractMatrix) = _eachslice(A, (2,), true)
+eachcol(A::AbstractVector) = eachcol(reshape(A, size(A,1), 1))
+
+"""
+    Rows{M,AX,S}
+
+A special case of [`Slices`](@ref) that is a vector of row slices of a matrix, as
+constructed by [`eachrow`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Rows{P<:AbstractMatrix,AX,S<:AbstractVector} = Slices{1,(1,:),P,AX,S}
+
+"""
+    Columns{M,AX,S}
+
+A special case of [`Slices`](@ref) that is a vector of column slices of a matrix, as
+constructed by [`eachcol`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Columns{P<:AbstractMatrix,AX,S<:AbstractVector} = Slices{1,(:,1),P,AX,S}
+
+
+IteratorSize(::Type{Slices{N,L,P,AX,S}}) where {N,L,P,AX,S} = HasShape{N}()
+axes(s::Slices) = s.axes
+size(s::Slices) = map(length, s.axes)
+
+@inline function _slice_index(s::Slices{N,L}, c...) where {N,L}
+    return map(l -> l === (:) ? (:) : c[l], L)
+end
+
+getindex(s::Slices, I...) = view(s.parent, _slice_index(s, I...)...)
```

Comment:
Here:

```suggestion
getindex(s::Slices{N}, I::Vararg{Int,N}...) where {N} = view(s.parent, _slice_index(s, I...)...)
```

This way Julia will ensure we always get `N` indices as `_slice_index` expects